Action fails when pinned by commit SHA (basename $GITHUB_ACTION_PATH not a release tag)

## Summary

When the action is pinned by commit SHA (e.g. `WillAbides/release-train@cd01ea9e19bb396bf659f61e1c840c5174b92de8`), the bootstrap script fails to fetch the binary because it interprets the SHA as a release tag.

Pinning third-party actions by commit SHA is GitHub's recommended security hardening practice ([docs](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions)), so this currently forces users to choose between SHA pinning and using release-train.

## Reproduction

```yaml
- uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
  with: { fetch-depth: 0 }
- uses: WillAbides/release-train@cd01ea9e19bb396bf659f61e1c840c5174b92de8 # v3.8.5
```

## Observed failure

```
Using release-train version cd01ea9e19bb396bf659f61e1c840c5174b92de8 from action
100     9  100     9    0     0     87      0 --:--:-- --:--:-- --:--:--    88
bindown: error: invalid config: jsonschema: '' does not validate with https://willabides.github.io/bindown/bindown.schema.json#/type: expected object, but got string
Successfully downloaded release-train cd01ea9e19bb396bf659f61e1c840c5174b92de8 from action
.../script/release-train: 36: exec: .../bin/release-train: not found
##[error]Process completed with exit code 127.
```

Real-world example: https://github.com/WillAbides/setup-go-faster/actions/runs/26668925342/job/78608017773?pr=41

## Root cause

In `script/release-train`:

```sh
ACTION_VERSION="$(basename "$GITHUB_ACTION_PATH")"
echo "Using release-train version $ACTION_VERSION from action"
…
curl -L -o tmp/bindown.yaml \
  "https://github.com/WillAbides/release-train/releases/download/$ACTION_VERSION/bindown.yaml"
```

When `uses:` references a SHA, GitHub clones the action to a path that ends with the SHA, so `basename "$GITHUB_ACTION_PATH"` returns the SHA — not the tag the SHA points at. The URL then 404s and curl writes the error page (9 bytes of HTML) into `tmp/bindown.yaml`.

## Three stacked failures

1. **Version derivation** — `basename "$GITHUB_ACTION_PATH"` is not a reliable source for the release tag. Should use a value baked into the action source (e.g. a `VERSION` file checked in alongside the script, or a constant interpolated from `action.yml`).

2. **`curl` doesn't fail on HTTP errors** — `curl -L -o file URL` happily writes a 404 body to the file. Adding `--fail` would surface the HTTP error and make the next step's failure self-explanatory.

3. **Bindown failure is masked** — bindown logs `error: invalid config: …` but the script still reaches `echo "Successfully downloaded release-train $ACTION_VERSION from action"` and then dies in `exec`. Either bindown is exiting 0 despite logging an error, or `set -e` isn't catching it inside the `( … )` subshell. The script claims success on the way to a fatal error.

## Suggested fix

Replace `basename "$GITHUB_ACTION_PATH"` with a release-tag value sourced from the action's own contents (e.g. write a one-line `VERSION` file during release and `cat` it in the script). Then `curl --fail --silent --show-error --location …` and `bindown … || { echo "bindown install failed"; exit 1; }` to make the two secondary failures loud if they ever recur.

## Workaround

For anyone hitting this who wants to keep SHA pinning today, the action's `release-train-bin` input lets you pre-download and pass the binary yourself:

```yaml
- name: Download release-train binary
  id: rt
  env:
    RT_VERSION: '3.8.5'
    RT_SHA256: '78d942e422f64089c92b54165023d64416b677b600244486579110f76a64c516'
  run: |
    set -euo pipefail
    url="https://github.com/WillAbides/release-train/releases/download/v${RT_VERSION}/release-train_${RT_VERSION}_linux_amd64.tar.gz"
    dest="$RUNNER_TEMP/release-train"
    mkdir -p "$dest"; cd "$dest"
    curl --fail --silent --show-error --location --output rt.tar.gz "$url"
    echo "${RT_SHA256}  rt.tar.gz" | sha256sum -c -
    tar -xzf rt.tar.gz
    chmod +x release-train
    echo "bin=$dest/release-train" >>"$GITHUB_OUTPUT"

- uses: WillAbides/release-train@<SHA>
  with:
    release-train-bin: ${{ steps.rt.outputs.bin }}
    # … other inputs
```

This is also a security improvement: the action source and the binary are independently pinned and verified.

--- a/script/release-train
+++ b/script/release-train
@@ -6,31 +6,113 @@ set -e
 
 REPO_DIR="$(CDPATH="" cd -- "$(dirname -- "$0")/.." && pwd -P)"
 
-action_dl() {
+CANONICAL_REPO="WillAbides/release-train"
+RELEASE_DOWNLOAD_BASE="https://github.com/$CANONICAL_REPO/releases/download"
+CANONICAL_GIT_URL="https://github.com/$CANONICAL_REPO.git"
+
+# looks_like_release_tag returns 0 if $1 looks like a semver release tag (vX.Y.Z...).
+looks_like_release_tag() {
+  case "$1" in
+    v[0-9]*.[0-9]*.[0-9]*) return 0 ;;
+  esac
+  return 1
+}
+
+# is_full_sha returns 0 if $1 is a 40-character hex string.
+is_full_sha() {
+  case "$1" in
+    *[!0-9a-fA-F]* | '') return 1 ;;
+  esac
+  [ "${#1}" -eq 40 ]
+}
+
+# resolve_sha_to_tag prints a semver release tag from the canonical repo that points at
+# the given commit SHA. Returns non-zero if ls-remote fails or no matching tag is found.
+resolve_sha_to_tag() {
+  _rstt_sha="$1"
+  _rstt_refs="$(git ls-remote --tags "$CANONICAL_GIT_URL" 2>/dev/null)" || return 1
+  _rstt_tag="$(printf '%s\n' "$_rstt_refs" | awk -v sha="$_rstt_sha" '
+    $1 == sha {
+      ref = $2
+      sub(/^refs\/tags\//, "", ref)
+      sub(/\^\{\}$/, "", ref)
+      if (ref ~ /^v[0-9]+\.[0-9]+\.[0-9]+/) {
+        print ref
+        exit
+      }
+    }
+  ')"
+  [ -n "$_rstt_tag" ] || return 1
+  printf '%s\n' "$_rstt_tag"
+}
+
+# determine_release_tag prints the release-train release tag that should be downloaded
+# based on the action's pinned ref, or returns non-zero to indicate that the script should
+# fall back to building from source.
+determine_release_tag() {
+  [ -n "$GITHUB_ACTION_PATH" ] || return 1
+  _drt_raw="$(basename "$GITHUB_ACTION_PATH")"
+
+  if looks_like_release_tag "$_drt_raw"; then
+    printf '%s\n' "$_drt_raw"
+    return 0
+  fi
+
+  # Only resolve a SHA when the action is pinned against the canonical repo. Forks at
+  # the same SHA can have tags pointing at unrelated source we cannot safely map to a
+  # canonical release.
+  _drt_repo="${GITHUB_ACTION_REPOSITORY:-$CANONICAL_REPO}"
+  [ "$_drt_repo" = "$CANONICAL_REPO" ] || return 1
+
+  is_full_sha "$_drt_raw" || return 1
+
+  _drt_tag="$(resolve_sha_to_tag "$_drt_raw")" || return 1
+  echo "Resolved action commit $_drt_raw to release tag $_drt_tag" >&2
+  printf '%s\n' "$_drt_tag"
+}
+
+# download_release downloads bindown.yaml for the given release tag and installs
+# release-train into bin/. Any failure here is fatal: the user explicitly pinned to a
+# released version, so silently falling back to a source build would mask the problem.
+download_release() {
   (
-    [ -n "$GITHUB_ACTION_PATH" ] || return 1
+    _dr_tag="$1"
     cd "$REPO_DIR"
-    ACTION_VERSION="$(basename "$GITHUB_ACTION_PATH")"
-    echo "Using release-train version $ACTION_VERSION from action"
+    echo "Using release-train version $_dr_tag from action"
 
-    mkdir -p tmp
-    rm -f tmp/bindown.yaml
+    mkdir -p tmp bin
+    rm -f tmp/bindown.yaml bin/release-train
 
-    curl -L -o tmp/bindown.yaml \
-      "https://github.com/WillAbides/release-train/releases/download/$ACTION_VERSION/bindown.yaml"
+    if ! curl --fail --silent --show-error --location \
+      --output tmp/bindown.yaml \
+      "$RELEASE_DOWNLOAD_BASE/$_dr_tag/bindown.yaml"; then
+      echo "release-train: failed to download bindown.yaml for $_dr_tag from $RELEASE_DOWNLOAD_BASE/$_dr_tag/bindown.yaml" >&2
+      return 1
+    fi
 
-    script/bindown install release-train --configfile tmp/bindown.yaml --cache .cache --output bin/release-train
-    echo "Successfully downloaded release-train $ACTION_VERSION from action" >&2
+    if ! script/bindown install release-train \
+      --configfile tmp/bindown.yaml --cache .cache --output bin/release-train; then
+      echo "release-train: failed to install release-train $_dr_tag via bindown" >&2
+      return 1
+    fi
+
+    echo "Successfully downloaded release-train $_dr_tag from action" >&2
   )
 }
 
 local_build() {
   (
     cd "$REPO_DIR"
+    mkdir -p bin
+    rm -f bin/release-train
     go build -o bin/release-train .
   )
 }
 
-action_dl || local_build
+if RELEASE_TAG="$(determine_release_tag)"; then
+  download_release "$RELEASE_TAG"
+else
+  local_build
+fi
 
 exec "$REPO_DIR"/bin/release-train "$@"

--- a/script/release-train
+++ b/script/release-train
@@ -30,7 +30,7 @@ is_full_sha() {
 # the given commit SHA. Returns non-zero if ls-remote fails or no matching tag is found.
 resolve_sha_to_tag() {
   _rstt_sha="$1"
-  _rstt_refs="$(git ls-remote --tags "$CANONICAL_GIT_URL" 2>/dev/null)" || return 1
+  _rstt_refs="$(git ls-remote --tags "$CANONICAL_GIT_URL" 2> /dev/null)" || return 1
   _rstt_tag="$(printf '%s\n' "$_rstt_refs" | awk -v sha="$_rstt_sha" '
     $1 == sha {
       ref = $2

